Expose file hashes as extended attributes.

eatnumber1 opened the pull request:
Hey there, in spite of the checklist below, I'd like some feedback on this direction before I put effort into making tests and implementing it for all the fuse backends. Can you take an early look?

#### What is the purpose of this change?

This change adds extended attributes found at `system.hash.<hash_type>`.
By reading these extended attributes, rclone will query and return the
hash from the storage backend without needing to download the file.

The following is an example of using these attributes when reading a
file stored in Google Drive using the Mac OSX `xattr` tool.

```
$ xattr image.jpg
system.hash.md5
$ xattr -p system.hash.md5 image.jpg
920e3d7899c8eed81b50d6f26fe0f64b
```

#### Was the change discussed in an issue or in the forum before?

Not directly, although there's been discussion about exposing hashes through the fs layer, and discussion about extended attributes in the past. The topics I've found are: #1710, #1190, and #3673.

#### Checklist

- [X] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [ ] I have added tests for all changes in this PR if appropriate.
- [ ] I have added documentation for the changes if appropriate.
- [X] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [ ] I'm done, this Pull Request is ready for review :-)

ncw commented:
I had a look through the code and it looks sensible to me :-)

What other options are there for extended attributes we could expose?

Are there any standards docs on extended attributes?

I'll get the CI to run on this change and see if it has any feedback.

eatnumber1 commented:
> Are there any standards docs on extended attributes?

Not that I can tell. From Linux's [xattr(7)](https://man7.org/linux/man-pages/man7/xattr.7.html):

```
CONFORMING TO
       Extended attributes are not specified in POSIX.1, but some other
       systems (e.g., the BSDs and Solaris) provide a similar feature.
```

With respect to documentation though, each platform has their own documentation. Here's the ones for [Linux](https://man7.org/linux/man-pages/man7/xattr.7.html), [OSX](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/getxattr.2.html), and [FreeBSD](https://www.freebsd.org/cgi/man.cgi?extattr). Supposedly there's also [equivalent functionality](https://en.wikipedia.org/wiki/Extended_file_attributes#Windows_NT) in Windows.

None of the platforms have completely compatible implementations, but the FUSE layer provides some amount of portability for FUSE filesystems. In the case of my code, I've tested it on OSX, and I'm fairly confident that it will work on Linux (in part because I'm conforming to Linux's namespacing rules), but I'll need to do a bit more research into how to get it working on BSD and especially Windows. If I'm lucky, it'll just work.

Another thing to consider here is my choice of exactly what key to use. I'm currently using `system.hash.<hash_type>`, where on Linux the `system` prefix is documented as follows:

```
  System extended attributes
      System extended attributes are used by the kernel to store system
      objects such as Access Control Lists.  Read and write access
      permissions to system attributes depend on the policy implemented
      for each system attribute implemented by filesystems in the
      kernel.
```

True we're not the kernel, but we're sort of a stand-in for the kernel here. I also don't think there's any actual enforcement that prevents us from using the system namespace at the FUSE layer, so it's somewhat of only a theoretical problem.

More practically though, there's other tools that store checksums as xattrs, so it might be useful to figure out if there's a de-facto standard for checksums-in-xattrs emerging, and reuse that. I did a bit of searching for this, and am not finding any consensus.

- [rmlint](https://rmlint.readthedocs.io/en/latest/) uses `user.rmlint.$USER.cksum`
- glusterfs [seems to use](https://bugzilla.redhat.com/show_bug.cgi?id=1208124) `trusted.glusterfs.bit-rot.signature`
- cshatag [uses](https://github.com/rfjakob/cshatag/blob/fe9aae114aa05987e8e54617dfeee088fb7f9bb7/check.go#L18) `user.shatag.sha256`
- various documentation about doing this manually seems to use [`user.checksum`](https://wiki.archlinux.org/title/File_permissions_and_attributes#User_extended_attributes) and [`user.checksum.md5`](https://www.linuxtoday.com/blog/extended-file-attributes-rock.html)

The Linux xattr documentation does give some examples, which includes an md5 hash:

```
  Extended attribute namespaces
      Attribute names are null-terminated strings.  The attribute name
      is always specified in the fully qualified namespace.attribute
      form, for example, user.mime_type, trusted.md5sum,
      system.posix_acl_access, or security.selinux.
```

So yeah, no clear consensus on what to use. Most tools seem to put their name in the key to prevent collisions, so there doesn't seem to be much attempt at reuse. Happy to do that for rclone too, or change it to `trusted.md5sum` to follow the example from the Linux docs (which I didn't notice until just now).

> What other options are there for extended attributes we could expose?

Hm, good question. For me my current goal is to get some ROM [management tooling](https://github.com/SmokeMonsterPacks/EverDrive-Packs-Lists-Database) to support organizing ROMs without downloading them (it manages them using the file's hashes), so I hadn't put much thought into other things we could export via xattrs.

Looking around a bit, I found https://www.freedesktop.org/wiki/CommonExtendedAttributes/ which lists both some useful design practices for naming xattrs, and some common ones in use today along with some proposed ones. Ones that seem useful for us include:

- `user.mime_type`
- `user.charset` (do we actually know the charset of files when applicable?)
- `user.xdg.origin.url` - perhaps put the URL from the backend here?

Looking at the Google Drive API [docs](https://developers.google.com/drive/api/v3/reference/files), there's a few more things that could be exported (although I don't know if it would actually be useful to do so), including

 - description
 - parents (since there can be more than one)
 - properties (this would be the read side of supporting arbitrary user specified xattrs)
 - version
 - thumbnailLink (I did some searching and it looks like thumbnails in most apps don't use xattrs, so proper thumbnail support would be a separate effort)
 - and a bunch more...

Then of course there would be the (mostly separate) work to allow user-defined xattrs.

> I'll get the CI to run on this change and see if it has any feedback.

Thanks! I'll finish up the other fuse backends, and write some tests and documentation in a bit.

eatnumber1 commented on cmd/cmount/fs.go:
```diff
@@ -550,7 +550,22 @@ func (fsys *FS) Setxattr(path string, name string, value []byte, flags int) (err
-	return -fuse.ENOSYS, nil
+	defer log.Trace(path, "name=%q", name)("errc=%d", &errc)
+
+	node, handle, errc := fsys.getNode(path, fhUnset)
```
Do you think it's worth grabbing the handle here considering we're passing `fhUnset` into `getNode`? From what I can tell, `getNode` will always return a `nil` handle in this case, so adding the code to cover the case where it doesn't would only be future-proofing if that behavior ever changes.

ncw replied:
If we don't have an `fh` then probably best to explicitly call this instead.

    node, errc := fsys.lookupNode(path)

eatnumber1 replied:
Done!

eatnumber1 commented:
Another thought: if we do eventually add support for user-defined xattrs, we could have special ones which set specific properties on the backend. For example, `user.gdrive.description` could set `description`, or `user.gdrive.starred` could set `starred`.

ncw reviewed (commented):
Can you check out the build failures in the CI - looks like there is more stuff to add!

ncw commented:
> > Are there any standards docs on extended attributes?
> 
> Not that I can tell. From Linux's [xattr(7)](https://man7.org/linux/man-pages/man7/xattr.7.html):
> 
> ```
> CONFORMING TO
>        Extended attributes are not specified in POSIX.1, but some other
>        systems (e.g., the BSDs and Solaris) provide a similar feature.
> ```
> 
> With respect to documentation though, each platform has their own documentation. Here's the ones for [Linux](https://man7.org/linux/man-pages/man7/xattr.7.html), [OSX](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/getxattr.2.html), and [FreeBSD](https://www.freebsd.org/cgi/man.cgi?extattr). Supposedly there's also [equivalent functionality](https://en.wikipedia.org/wiki/Extended_file_attributes#Windows_NT) in Windows.
> 
> None of the platforms have completely compatible implementations, but the FUSE layer provides some amount of portability for FUSE filesystems. In the case of my code, I've tested it on OSX, and I'm fairly confident that it will work on Linux (in part because I'm conforming to Linux's namespacing rules), but I'll need to do a bit more research into how to get it working on BSD and especially Windows. If I'm lucky, it'll just work.

Cross fingers! I'm not sure how WinFSP deals with attributes.

> Another thing to consider here is my choice of exactly what key to use. I'm currently using `system.hash.<hash_type>`, where on Linux the `system` prefix is documented as follows:
> 
> ```
>   System extended attributes
>       System extended attributes are used by the kernel to store system
>       objects such as Access Control Lists.  Read and write access
>       permissions to system attributes depend on the policy implemented
>       for each system attribute implemented by filesystems in the
>       kernel.
> ```
> True we're not the kernel, but we're sort of a stand-in for the kernel here. I also don't think there's any actual enforcement that prevents us from using the system namespace at the FUSE layer, so it's somewhat of only a theoretical problem.

Yes I would say `system` is appropriate here. As a FUSE fs we are standing in for the kernel. `user` prefix is for random stuff the user wants to store.

> More practically though, there's other tools that store checksums as xattrs, so it might be useful to figure out if there's a de-facto standard for checksums-in-xattrs emerging, and reuse that.

Good idea

> I did a bit of searching for this, and am not finding any consensus.
> 
> * [rmlint](https://rmlint.readthedocs.io/en/latest/) uses `user.rmlint.$USER.cksum`
> * glusterfs [seems to use](https://bugzilla.redhat.com/show_bug.cgi?id=1208124) `trusted.glusterfs.bit-rot.signature`
> * cshatag [uses](https://github.com/rfjakob/cshatag/blob/fe9aae114aa05987e8e54617dfeee088fb7f9bb7/check.go#L18) `user.shatag.sha256`
> * various documentation about doing this manually seems to use [`user.checksum`](https://wiki.archlinux.org/title/File_permissions_and_attributes#User_extended_attributes) and [`user.checksum.md5`](https://www.linuxtoday.com/blog/extended-file-attributes-rock.html)
> 
> The Linux xattr documentation does give some examples, which includes an md5 hash:
> 
> ```
>   Extended attribute namespaces
>       Attribute names are null-terminated strings.  The attribute name
>       is always specified in the fully qualified namespace.attribute
>       form, for example, user.mime_type, trusted.md5sum,
>       system.posix_acl_access, or security.selinux.
> ```
> 
> So yeah, no clear consensus on what to use. Most tools seem to put their name in the key to prevent collisions, so there doesn't seem to be much attempt at reuse. Happy to do that for rclone too, or change it to `trusted.md5sum` to follow the example from the Linux docs (which I didn't notice until just now).

I guess `rclone.md5sum` is absolutely explicit and won't step on anyone's toes? I'd rather use a standard name if we can figure out the best one though.

> > What other options are there for extended attributes we could expose?
> 
> Hm, good question. For me my current goal is to get some ROM [management tooling](https://github.com/SmokeMonsterPacks/EverDrive-Packs-Lists-Database) to support organizing ROMs without downloading them (it manages them using the file's hashes), so I hadn't put much thought into other things we could export via xattrs.
> 
> Looking around a bit, I found https://www.freedesktop.org/wiki/CommonExtendedAttributes/ which lists both some useful design practices for naming xattrs, and some common ones in use today along with some proposed ones. Ones that seem useful for us include:
> 
> * `user.mime_type`

This would be easy to add. You can find the mime type of any `Object` with the `fs.MimeType()` call.

> * `user.charset` (do we actually know the charset of files when applicable?)

No.

> * `user.xdg.origin.url` - perhaps put the URL from the backend here?

This could potentially be the output of `rclone link` (the internal method is called PublicLink).

I think maybe we should add `user.mime_type` as well 1) because it is useful, and 2) to generalise the very special case system we have at the moment.

> Looking at the Google Drive API [docs](https://developers.google.com/drive/api/v3/reference/files), there's a few more things that could be exported (although I don't know if it would actually be useful to do so), including
> 
> * description
> * parents (since there can be more than one)
> * properties (this would be the read side of supporting arbitrary user specified xattrs)
> * version
> * thumbnailLink (I did some searching and it looks like thumbnails in most apps don't use xattrs, so proper thumbnail support would be a separate effort)
> * and a bunch more...

Do be useful they would need to exposed via the Object layer which they aren't right now.

The format specifiers for `rclone lsf` give an idea of things which would be easy to export.

```
p - path
s - size
t - modification time
h - hash
i - ID of object
o - Original ID of underlying object
m - MimeType of object if known
e - encrypted name
T - tier of storage if known, e.g. "Hot" or "Cool"
```

> Then of course there would be the (mostly separate) work to allow user-defined xattrs.

This would need support from the backend to store the xattrs which is part of the general metadata for each backend project.

> > I'll get the CI to run on this change and see if it has any feedback.

It did!

> Another thought: if we do eventually add support for user-defined xattrs, we could have special ones which set specific properties on the backend. For example, `user.gdrive.description` could set `description`, or `user.gdrive.starred` could set `starred`.

That is possible certainly...

What I envision eventually is that each backend exposes a set of metadata, some of which will be backend specific (eg `starred` and some of which may be set by the user). This would then be exposed as xattrs and when copying from backend to backend "compatible" metadata would be copied, with the rest being stored as user keys.

eatnumber1 commented:
Okay I've fleshed this out a bit. Support is in place for both the cmount and mount2 backends, and some (very simplistic) test are added to the vfs layer. In addition, I added the user.mime_type attribute.

Please take a look at the commit messages, as I made a few changes.

As for what attribute to put the hashes under, I ended up settling for `system.org.rclone.hash` to follow the freedesktop.org recommendations. Happy to change it if you'd prefer something else though.

I've actually also implemented support in the mount backend, but I wasn't able to test it (I wasn't even able to compile it) as I'm testing on OSX. For that reason, the changes to support the mount backend aren't part of this PR but can be seen [here](https://github.com/eatnumber1/rclone/commit/d68f1dd5c98e8c3f5c98c579d6d04eca3fcbcf7c). What, if anything, would you like me to do with it?

> Can you check out the build failures in the CI - looks like there is more stuff to add!

I moved the xattr logic out of the Handle (since we don't actually need an open file to read xattrs) which fixed it. I tried replicating the CI a bit with the following:

```
$ RCLONE_CONFIG="/notfound" go test -tags "cmount" ./...
```

... and that passed on my mac.

So if the CI runs again and is happy with my changes, I think the only remaining thing is to add documentation. Do you have a suggestion on the best place to put docs on this feature?

Also to be clear, I've tested this with both cmount and mount2 on my mac, but I haven't tested this on Linux, BSD, or Windows.

eatnumber1 commented on cmd/mount2/node.go:
```diff
@@ -404,3 +404,83 @@ func (n *Node) Rename(ctx context.Context, oldName string, newParent fusefs.Inod
+		return
+	}
+
+	copy(dest, data)
```
If we change the API of n.node.Getxattr to take a non-reallocateable slice, it moves this complexity into the vfs layer, but is more performant (we can skip this copy). Do you have a preference on simplicity vs. performance in the vfs layer?

eatnumber1 commented on vfs/file.go:
```diff
@@ -784,3 +787,46 @@ func (f *File) Truncate(size int64) (err error) {
+	obj := f.getObject()
+
+	if name == "user.mime_type" {
+		return []byte(fs.MimeType(context.TODO(), obj)), nil
```
In the mount2 backend it does give us Context objects that could be passed up. I see however that none of these other methods accept a Context, so I didn't add it to this one.

Do you think it's worth adding a Context argument for Getattr/Listattr?

eatnumber1 replied:
Opted to pass the context object through.

eatnumber1 commented:
Ah, based on the CI failures I'm reminded that there's disagreement between platforms on whether ENOATTR exists. I'll fix that shortly.

eatnumber1 commented:
I've:

1. Fixed the ENOATTR issue
2. Plumbed the context through from the mount layer.
3. Tested on Linux
4. Implemented the mount backend (and tested it on Linux)

As mentioned all three mount backends work correctly on Linux, although it looks like `getfattr` hides the system namespace by default, so you need to pass `-m '.*'` to get the hash to appear.

Here's my tests, with identical output for all three mount backends:

```
$ getfattr -d -m '.*' image.jpg
getfattr: Removing leading '/' from absolute path names
# file: image.jpg
system.org.rclone.hash.md5="920e3d7899c8eed81b50d6f26fe0f64b"
user.mime_type="image/jpeg"
$ getfattr -n user.noexist image.jpg
image.jpg: user.noexist: No such attribute
```

eatnumber1 commented:
Fixed a couple of style issues, which should fix the CI for Linux.

For FreeBSD, I had to make a few fixes to the way I was handling ENODATA. That's done now, but I don't have a FreeBSD machine to test, so may need a few rounds with the CI to get it right.

eatnumber1 commented:
I just had a thought: it would be nice if file hashes get preserved on files even if you copy the file out of rclone. Putting the hash in the system namespace won't allow that to happen (on Linux).

I'm gonna change it to `user.checksum.<hashtype>` so the hashes can be preserved.

eatnumber1 commented:
Fixed missing syscall import on Linux and FreeBSD, and manually tested on FreeBSD (which I hadn't done before).

```
$ lsextattr user image.jpg
image.jpg	checksum.md5	mime_type
$ lsextattr user image.jpg
image.jpg	checksum.md5	mime_type
$ getextattr user checksum.md5 image.jpg
image.jpg	920e3d7899c8eed81b50d6f26fe0f64b
$ getextattr user mime_type image.jpg
image.jpg	image/jpeg
```

eatnumber1 commented:
Is the test failure on Linux a flaky test? It seems to pass on my Linux machine:

```
go test -count=10 -tags=cmount,mount2,mount github.com/rclone/rclone/cmd/mount2
ok      github.com/rclone/rclone/cmd/mount2     58.844s
```

ncw commented:
> Is the test failure on Linux a flaky test? It seems to pass on my Linux machine:
> 
> ```
> go test -count=10 -tags=cmount,mount2,mount github.com/rclone/rclone/cmd/mount2
> ok      github.com/rclone/rclone/cmd/mount2     58.844s
> ```

Hmm, I think it is probably a flakey test.

I'll kick the CI to have another go.

ncw commented:
Can you squash the noise commits (reformatting, fix this etc) into the relevant commit so I have a nice clean set of commits and then I'll review them one more time :-)

Thanks

eatnumber1 commented:
Cleaned up most of them except for c61af39b454ab84478bb297ef0e85cc4753d3fdc, which will require some merge deconflicting to eliminate, so I left it. Let me know if you'd prefer I remove it.

I opted to leave the commit history of the evolution from `system.hash` through `system.org.rclone.hash` to `user.checksum` just in case it's useful for others to see why we decided on `user.checksum` .

eatnumber1 commented:
According to the CI, this is the errors:

```
cmd/cmount/errors.go:1: File is not `goimports`-ed (goimports)
// +build cmount
cmd/mount/file.go:135: File is not `goimports`-ed (goimports)
		return true  // keep going
```

I've fixed them both.

Running golangci-lint manually, I get similar errors from across rclone though. It looks like the `// +build` syntax [has changed](https://pkg.go.dev/go/build#hdr-Build_Constraints) to `//go:build`, which is generating these errors. I suppose the CI isn't complaining about files I didn't change?

ivandeex commented:
you can either
- install go 1.17 and run `gofmt -w ./file1.go ...; git commit ...`
- or analyze the latest build finished after this very comment

eatnumber1 commented:
```
Error: cmd/all/all.go:15:2: //go:build comment without // +build comment
```

Fixed.

I'll see about installing go 1.17 (I'm on 1.13) so I don't keep needing you to rerun the CI. Sorry for the trouble.

eatnumber1 commented:
I installed go 1.17 and indeed, gofmt now reports warnings for most of the files in rclone (all those with `// +build` comments).

```
$ gofmt -l . | wc -l
177
```

I've now updated the build comment for all of the files I touched.

```
gofmt -w $(git log fb1b45086593fecd6e765dbbec8802580649f4e6..HEAD | grep commit | cut -d\  -f2 | xargs -n1 git diff-tree --name-only --no-commit-id -r | sort -u | grep '.*\.go$')
```

eatnumber1 commented:
Friendly ping here, looks like because this has been sitting for a while it needs some rebasing to merge successfully now. I'm happy to do that if you'd prefer, but would like to get this submitted.

eatnumber1 commented:
Is there anything left for me to do here, or are we just waiting for @ncw to review?

eatnumber1 commented:
@ncw friendly ping, are you still willing to pull in this change?

ivandeex commented:
We are waiting for me. I'm busy elsewhere now. Will review outstanding PRs assigned on me in a month.
Sorry for delay.

eatnumber1 commented:
Hey there, if I do the necessary rebasing, is this still a PR that can get merged?

ncw commented:
Firstly, many apologies for the long delay on this. So many pull requests - so little time!

There have been some developments in the rclone code base since you did this, namely [the metadata framework](https://rclone.org/docs/#metadata) and the [local backend's support of metadata](https://rclone.org/local/#metadata) so what I'd suggest would be the best way forward would be for this PR to expose the metadata on objects in the same way the local backend does, so on a `user.*` prefix.

There is more work to be done on the metadata framework (its only supported by s3 and local at the moment) but doing it like this gives proper support for all backends.

Backends don't currently expose hashes as metadata - maybe they should.

What do you think?
